### PR TITLE
wallet: fix when sufficient preset inputs and subtractFeeFromOutputs

### PR DESCRIPTION
**Pull request by achow101**
#17290 introduced a bug where, when we had preset inputs that covered the amount being sent and subtractFeeFrromOutputs was being used, transaction funding would result in a `Fee exceeds maximum configured by -maxtxfee` error. This was happening because we weren't setting `bnb_used = false` when the preset inputs were used as it should have been. This resulted in a too high fee because the change would go to fees accidentally.

Apparently this particular case doesn't have a test, so I've added one as well.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2349,14 +2349,19 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
     size_t max_descendants = (size_t)std::max<int64_t>(1, limit_descendant_count);
     bool fRejectLongChains = gArgs.GetBoolArg("-walletrejectlongchains", DEFAULT_WALLET_REJECT_LONG_CHAINS);
 
-    bool res = value_to_select <= 0 ||
+    bool res = value_to_select <= 0;
+    if (res) {
+        bnb_used = false;
+    } else {
+        res = value_to_select <= 0 ||
```
Comment (meshcollider) [inline src/wallet/wallet.cpp]: Why leave this case here when we know it is false?

Comment (achow101) [reply src/wallet/wallet.cpp]: Oops, forgot to delete that.

**Comment by meshcollider** [comment]
The new test passes on current master without the code change

**Comment by achow101** [comment]
> The new test passes on current master without the code change

Not for me:

```
2019-11-23T05:27:14.580000Z TestFramework (INFO): Test fundrawtxn subtract fee from outputs with preset inputs that are sufficient
2019-11-23T05:27:14.637000Z TestFramework (ERROR): JSONRPC error
Traceback (most recent call last):
  File "/home/andy/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 111, in main
    self.run_test()
  File "test/functional/rpc_fundrawtransaction.py", line 95, in run_test
    self.test_subtract_fee_with_presets()
  File "test/functional/rpc_fundrawtransaction.py", line 753, in test_subtract_fee_with_presets
    fundedtx = self.nodes[0].fundrawtransaction(rawtx, {'subtractFeeFromOutputs': [0]})
  File "/home/andy/bitcoin/bitcoin/test/functional/test_framework/coverage.py", line 47, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/andy/bitcoin/bitcoin/test/functional/test_framework/authproxy.py", line 141, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: Fee exceeds maximum configured by -maxtxfee (-4)
2019-11-23T05:27:14.694000Z TestFramework (INFO): Stopping nodes
2019-11-23T05:27:14.849000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_rpwug9f9
2019-11-23T05:27:14.849000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_rpwug9f9/test_framework.log
2019-11-23T05:27:14.849000Z TestFramework (ERROR): Hint: Call /home/andy/bitcoin/bitcoin/test/functional/combine_logs.py '/tmp/bitcoin_func_test_rpwug9f9' to consolidate all logs
```

**Comment by laanwj** [comment]
I've tested this.
When running the new test on current master's binary, the "rpc_fundrawtransaction.py" test gives me a "test_framework.authproxy.JSONRPCException: Fee exceeds maximum configured by -maxtxfee (-4)" consistently.
After recompile with this patch, it passes.

This is really strange, I wonder what causes the difference in behavior.

**Comment by darosior** [comment]
FWIW the test pass on master for me as well (without the code change).

**Comment by meshcollider** [comment]
```
samuel@hansome-pc:~/git/bitcoin$ git diff master
diff --git a/test/functional/rpc_fundrawtransaction.py b/test/functional/rpc_fundrawtransaction.py
index 693051edc..6f1ae0d3b 100755
--- a/test/functional/rpc_fundrawtransaction.py
+++ b/test/functional/rpc_fundrawtransaction.py
@@ -92,6 +92,7 @@ class RawTransactionsTest(BitcoinTestFramework):
         self.test_option_feerate()
         self.test_address_reuse()
         self.test_option_subtract_fee_from_outputs()
+        self.test_subtract_fee_with_presets()

     def test_change_position(self):
         """Ensure setting changePosition in fundraw with an exact match is handled properly."""
@@ -741,5 +742,17 @@ class RawTransactionsTest(BitcoinTestFramework):
         # The total subtracted from the outputs is equal to the fee.
         assert_equal(share[0] + share[2] + share[3], result[0]['fee'])

+    def test_subtract_fee_with_presets(self):
+        self.log.info("Test fundrawtxn subtract fee from outputs with preset inputs that are sufficient")
+
+        addr = self.nodes[0].getnewaddress()
+        txid = self.nodes[0].sendtoaddress(addr, 10)
+        vout = find_vout_for_address(self.nodes[0], txid, addr)
+
+        rawtx = self.nodes[0].createrawtransaction([{'txid': txid, 'vout': vout}], [{self.nodes[0].getnewaddress(): 5}])
+        fundedtx = self.nodes[0].fundrawtransaction(rawtx, {'subtractFeeFromOutputs': [0]})
+        signedtx = self.nodes[0].signrawtransactionwithwallet(fundedtx['hex'])
+        self.nodes[0].sendrawtransaction(signedtx['hex'])
+
 if __name__ == '__main__':
     RawTransactionsTest().main()
samuel@hansome-pc:~/git/bitcoin$ make -j5
Making all in src
make[1]: Entering directory '/home/samuel/git/bitcoin/src'
make[2]: Entering directory '/home/samuel/git/bitcoin/src'
  CXX      libbitcoinconsensus_la-arith_uint256.lo
  CXX      libbitcoinconsensus_la-hash.lo
  CXX      libbitcoinconsensus_la-pubkey.lo
  CXX      libbitcoinconsensus_la-uint256.lo
...  
  CXX      qt/qt_libbitcoinqt_a-moc_walletview.o
  CXX      qt/qt_libbitcoinqt_a-qrc_bitcoin.o
  CXX      qt/qt_libbitcoinqt_a-qrc_bitcoin_locale.o
  CXXLD    bitcoind
  CXXLD    test/test_bitcoin
  CXXLD    bench/bench_bitcoin
  AR       qt/libbitcoinqt.a
  CXXLD    qt/bitcoin-qt
  CXXLD    qt/test/test_bitcoin-qt
make[2]: Leaving directory '/home/samuel/git/bitcoin/src'
make[1]: Leaving directory '/home/samuel/git/bitcoin/src'
Making all in doc/man
make[1]: Entering directory '/home/samuel/git/bitcoin/doc/man'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/samuel/git/bitcoin/doc/man'
make[1]: Entering directory '/home/samuel/git/bitcoin'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/samuel/git/bitcoin'
samuel@hansome-pc:~/git/bitcoin$ ./test/functional/test_runner.py rpc_fundrawtransaction.py
Temporary test directory at /tmp/test_runner_?_??_20191124_180027
Remaining jobs: [rpc_fundrawtransaction.py]
1/1 - rpc_fundrawtransaction.py passed, Duration: 13 s

TEST                      | STATUS    | DURATION

rpc_fundrawtransaction.py | ? Passed  | 13 s

ALL                       | ? Passed  | 13 s (accumulated)
Runtime: 13 s
```

Clean branch, up to date master, fresh build. Idk why it passes but yours doesn't

**Comment by achow101** [comment]
I figured out why we are getting different test results on master.

It's because bnb_used is not initialized, so the access is undefined behavior and changes depending on the optimizations used. I always configured with `--enable-debug` which disables all optimizations. Presumably this results in `bnb_used` having some value that corresponds to `true`. After configuring without `--enable-debug` and thus allowing optimizations, the test passes on master. I assume this is because some optimization gives `bnb_used` the `false` value by default.

***

I've pushed another commit which should completely address this type of issue, at least until we merge #17331 which gets rid of that variable entirely. Instead of setting it at each of the non-BnB cases that can occur, we just default assume `bnb_used = false` and let it be set to `true` after BnB selection succeeds.

**Comment by meshcollider** [comment]
Code review ACK d3ca0b4a9621bdab47d22229adbf6f81c0d11384

**Review by darosior — APPROVED** [review]
ACK 4b0d19babf20291fc8c3dadad106367b06446e9d

**Comment by Sjors** [comment]
@practicalswift spotted a similar issue in #13546. Let's also actually initialize this thing, as @promag [suggested](https://github.com/bitcoin/bitcoin/pull/13546#issuecomment-400985313):
https://github.com/bitcoin/bitcoin/blob/2a97d2b1a525dbc8f6e6610fe543e003acfb7fac/src/wallet/wallet.cpp#L2681

**Review by instagibbs — COMMENTED** [review]
Second commit needs a rework of the commit message since it's just an additional test now.

**Comment by achow101** [comment]
Changed `bnb_used` to be initialized to false. Also updated the commit message.

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/17568/commits/eadd1304c81e0b89178e4cc7630bd31650850c85

**Comment by Sjors** [comment]
ACK eadd130. I can't get this new test to fail on macOS (without this PR). It passes whether or not I compile with `--enable-debug`. It does fail on Ubuntu. Yay undefined behavior... Anyway, it's a useful test.

**Review by fanquake — APPROVED** [review]
ACK eadd1304c81e0b89178e4cc7630bd31650850c85

`SelectCoins()` is only called in `CreateTransaction()`. `bnb_used` is set to false before that call (and again inside `SelectCoins()`) which is before any calls to `SelectCoinsMinConf()`.

Looking forward to something like #17331 which not only removes `bnb_used` but simplifies the wallet.cpp logic. As we've currently got code like this:
```cpp
// Choose coins to use
bool bnb_used = false;
if (pick_new_inputs) {
    <snip>
} else {
    bnb_used = false;
}
```
and are doing similar things with `coin_selection_params.use_bnb`.

**Comment by promag** [comment]
@Sjors thanks for the mention.

Code review ACK eadd1304c81e0b89178e4cc7630bd31650850c85.

**Pull request merged**